urlc5fdb179cc_lagodiuk_evo_neural_network_agents_tgz-pJ8-com_lagodiuk_agent_evolution_MainJ8/src/com/lagodiuk/agent/evolution/NeuralNetworkDrivenAgent.java

```diff
@@ -0,0 +1,255 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ ******************************************************************************/
+package com.lagodiuk.agent.evolution;
+
+import java.util.LinkedList;
+import java.util.List;
+
+import com.lagodiuk.agent.AbstractAgent;
+import com.lagodiuk.agent.Agent;
+import com.lagodiuk.agent.AgentsEnvironment;
+import com.lagodiuk.agent.Food;
+import com.lagodiuk.nn.NeuralNetwork;
+import com.lagodiuk.nn.ThresholdFunction;
+import com.lagodiuk.nn.genetic.OptimizableNeuralNetwork;
+
+public class NeuralNetworkDrivenAgent extends Agent {
+
+	private static final double maxSpeed = 4;
+
+	private static final double maxDeltaAngle = 1;
+
+	protected static final double maxAgentsDistance = 5;
+
+	private static final double AGENT = -10;
+
+	private static final double EMPTY = 0;
+
+	private static final double FOOD = 10;
+
+	private volatile NeuralNetwork brain;
+
+	public NeuralNetworkDrivenAgent(double x, double y, double angle) {
+		super(x, y, angle);
+	}
+
+	/**
+	 * Animating of agents and evolving best brain - might be in different
+	 * threads <br/>
+	 * Synchronization prevents from race condition when trying to set new
+	 * brain, while method "interact" runs <br/>
+	 * <br/>
+	 * TODO Maybe consider to use non-blocking technique. But at the moment this
+	 * simplest solution doesn't cause any overheads
+	 */
+	public synchronized void setBrain(NeuralNetwork brain) {
+		this.brain = brain;
+	}
+
+	/**
+	 * Synchronization prevents from race condition when trying to set new
+	 * brain, while method "interact" runs <br/>
+	 * <br/>
+	 * TODO Maybe consider to use non-blocking technique. But at the moment this
+	 * simplest solution doesn't cause any overheads
+	 */
+	@Override
+	public synchronized void interact(AgentsEnvironment env) {
+		List<Double> nnInputs = this.createNnInputs(env);
+
+		this.activateNeuralNetwork(nnInputs);
+
+		int neuronsCount = this.brain.getNeuronsCount();
+		double deltaAngle = this.brain.getAfterActivationSignal(neuronsCount - 2);
+		double deltaSpeed = this.brain.getAfterActivationSignal(neuronsCount - 1);
+
+		deltaSpeed = this.avoidNaNAndInfinity(deltaSpeed);
+		deltaAngle = this.avoidNaNAndInfinity(deltaAngle);
+
+		double newSpeed = this.normalizeSpeed(this.getSpeed() + deltaSpeed);
+		double newAngle = this.getAngle() + this.normalizeDeltaAngle(deltaAngle);
+
+		this.setAngle(newAngle);
+		this.setSpeed(newSpeed);
+
+		this.move();
+	}
+
+	private double avoidNaNAndInfinity(double x) {
+		if ((Double.isNaN(x)) || Double.isInfinite(x)) {
+			x = 0;
+		}
+		return x;
+	}
+
+	private void activateNeuralNetwork(List<Double> nnInputs) {
+		for (int i = 0; i < nnInputs.size(); i++) {
+			this.brain.putSignalToNeuron(i, nnInputs.get(i));
+		}
+		this.brain.activate();
+	}
+
+	protected List<Double> createNnInputs(AgentsEnvironment environment) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 11. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 144-180 into a separate method.